package/boot/arm-trusted-firmware-mvebu/Makefile

```diff
@@ -25,29 +25,70 @@ include $(INCLUDE_DIR)/package.mk
 define Package/arm-trusted-firmware-mvebu
   SECTION:=boot
   CATEGORY:=Boot Loaders
-  TITLE:=ARM Trusted Firmware for mvebu devices
+  DEPENDS:=@TARGET_mvebu_cortexa53
+  VARIANT:=$(subst arm-trusted-firmware-mvebu-,,$(1))
 endef
 
-define Package/arm-trusted-firmware-mvebu/uDPU
-$(call Package/arm-trusted-firmware-mvebu)
-  BOARDNAME:=uDPU
+define Package/arm-trusted-firmware-mvebu-espressobin-512mb
+  $(Package/arm-trusted-firmware-mvebu)
+  TITLE:=ARM Trusted Firmware for Marvell ESPRESSObin (512MB)
+  DEPENDS:=+u-boot-mvebu_espressobin-88f3720
+  UBOOT:=mvebu_espressobin-88f3720
+  DDR_TOPOLOGY:=0
+  CLOCKSPRESET:=CPU_800_DDR_800
```

Comment:
Why are You lowering the CPU clock?

Reply:
It's not lowered, there was a bug in kernel < 1.14, 800/800 is equal to what shipped as 100/800 with those old kernel versions. 800/800 is also the only stable version, there're multiple reports on armbian's forum that anything faster is unstable.

Reply:
I added the 800/800 explanation to the commit message.

Reply:
Ok, let's keep that.

Reply:
@ttocsr Hi, can You show output of this tool: https://github.com/wtarreau/mhz on uDPU?

Reply:
@tmn505 
count=330570 us50=20942 us250=104846 diff=83904 cpu_MHz=787.972

Reply:
@ttocsr similar numbers as ESPRESSObin with 1000_800 clockpreset. Well we'll have to live with it. Thanks.

Reply:
Just to note, on Espressobin v5 with shipped u-boot and Debian stable kernel I'm getting:
count=413212 us50=20691 us250=103508 diff=82817 cpu_MHz=997.892

Reply:
@pali what's the exact U-Boot and kernel version?

Reply:
I do not have access to serial console to that device, just ssh.
```
$ uname -a
Linux Espressobin 4.19.0-10-arm64 #1 SMP Debian 4.19.132-1 (2020-07-24) aarch64 GNU/Linux
```
And via `strings` on /dev/mtd* I see these lines:
```
-armada-17.10.1-b90dbf0
v1.3(release):armada-17.10.3:a3306ab
Built : 18:22:43, Jan 29 2018
v1.3(release):armada-17.10.3:a3306ab
Built : 18:22:44, Jan 29 2018
v1.3(release):armada-17.10.3:a3306ab
Built : 18:22:47, Jan 29 2018
U-Boot 2017.03-armada-17.10.1-gaee49fc (Jan 29 2018 - 18:21:49 +0800)
mvebu_armada-37xx
2017.03-armada-17.10.1-gaee49fc
01/29/2018
```
It is enough?

Reply:
Thanks, that's enough.

Reply:
I think it has to do with cpu frequency scaling.  From looking on Armbian forums.

Reply:
I'll check tomorrow if this has something to do with that.

Reply:
With a firmware built from this PR I get this on my v5 espressobin:
count=330570 us50=20881 us250=104060 diff=83179 cpu_MHz=794.840

Reply:
On another espressobin v5 I compiled upstream u-boot (with patches) and atf-marvell from atf-v1.5-armada-18.12 branch (the most recent one) with option CLOCKSPRESET=CPU_1000_DDR_800 and mhz prints:
count=330570 us50=20774 us250=103848 diff=83074 cpu_MHz=795.845

Reply:
On that same (second) espressobin v5 as in previous comment, I compiled upstream u-boot (with same patch set) but with upstream ATF (not marvell fork) with option CLOCKSPRESET=CPU_1000_DDR_800 and mhz prints:
count=330570 us50=20774 us250=103862 diff=83088 cpu_MHz=795.711

So it looks like that CLOCKSPRESET=CPU_1000_DDR_800 is broken also in upstream ATF and locked to 800MHz.

Interesting is that on first espressobin v5 on which is running factory firmware/u-boot is 1000MHz.

@tmn505 any idea where is the problem or why it is locked also with upstream ATF? And why original factory image from marvell/globalscale is not locked at all?

Reply:
On second espressobin v5 I replaced kernel by 5.8 version and with upstream u-boot and upstream ATF and CLOCKSPRESET=CPU_1000_DDR_800 I'm getting:
count=413212 us50=20691 us250=103473 diff=82782 cpu_MHz=998.314

Reply:
Got delayed a day, sorry.
So as @ttocsr mentioned the issue depends on which scaling governor one is using. If the governor is set to 'ondemand' the mhz tool can't report above 800MHz, even if more iterations are specified. On the other hand, if set to 'performance' it goes close to 1000 MHz on first try. I think either the tool has an issue or there's other problem.

Reply:
I guess it's the tool?
With a CPU_1000_DDR_800 build from this PR, kernel 5.6.14-2~bpo10+1, and the schedutil govenor I too only get cpu_MHz=793.876.
But cpufreq-info reports hardware limits: 200 MHz - 1000 MHz

Reply:
Debian has by default `performance` governor so it seems to be plausible. On first espressobin (where is Debian) armada-37xx-cpufreq was not automatically loaded so cpufreq was not in use. When I loaded it, nothing changed and I still get value around 1000MHz. After I changed governor to `ondemand` mhz started reporting around 790MHz.

Reply:
I did more testing on Turris MOX (which have also A3720 1GHz CPU) and `ondemand` governor is really problem.

By default `performance` governor is loaded and mhz reports number around 1000 MHz. After I switched governor to `ondemand` mhz started reporting freq to around 800 MHz. And 800 MHz stays even after switching governor back to `performance`.

So it looks like bug is in armada-37xx-cpufreq.c driver.

Reply:
There is really bug in armada-37xx-cpufreq and armada-37xx-periph drivers. Marek has already prepared patches.

@tmn505 @dhewg @ttocsr Would you like to test Marek's patches for this cpufreq issue?

Reply:
I'm not sure I'll find the time to test that soonish, but please link to it and I'll try!

Reply:
Patches for fixing cpufreq are there:
https://lore.kernel.org/linux-arm-kernel/20201007193522.9265-1-kabel@kernel.org/T/#u

But we found out that changing frequency from 500MHz to 1000MHz is unstable and CPU in more cases crashes. It looks like Marvell knows about this problem as they provided ["hack"](https://github.com/MarvellEmbeddedProcessors/linux-marvell/commit/dc33b62c90696afb6adc7dbcc4ebbd48bedec269) for old Espressobin V7 kernel to make it stable. This hack is not in upstream kernel and probably is the reason for instability of 1000MHz and 1200MHz variants of Espressobin V7.

Here is that Marvell hack rewritten for upstream kernel:
```patch
diff --git a/drivers/cpufreq/armada-37xx-cpufreq.c b/drivers/cpufreq/armada-37xx-cpufreq.c
index 4ee6b41e92a5..f664b78b30f7 100644
--- a/drivers/cpufreq/armada-37xx-cpufreq.c
+++ b/drivers/cpufreq/armada-37xx-cpufreq.c
@@ -258,6 +258,8 @@ static void __init armada37xx_cpufreq_avs_configure(struct regmap *base,
 	target_vm = avs_map[l0_vdd_min] - 150;
 	target_vm = target_vm > MIN_VOLT_MV ? target_vm : MIN_VOLT_MV;
 	dvfs->avs[2] = dvfs->avs[3] = armada_37xx_avs_val_match(target_vm);
+
+	dvfs->avs[1] = dvfs->avs[2] = dvfs->avs[3] = armada_37xx_avs_val_match(1108);
 }
 
 static void __init armada37xx_cpufreq_avs_setup(struct regmap *base,
```
Without it, Turris MOX (1GHz A3720) with ondemand governor is crashing.